File: buy_and_sell_stock.py
#time: O(n)
#space:O(1)
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices)<2:
            return 0
        max_val=min_val=prices[0]
        profit = 0
        
        for val in (prices[1:]):
            if val>=max_val:
                max_val=val
                # profit+= max(profit,max_val-min_val)
            else:
                profit+= max_val-min_val
                min_val=max_val=val
        profit+= max_val-min_val
        return profit

File: buyAndSellStock.java
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None

public class buyAndSellStock {
    public int maxProfit(int[] prices) {
        if(prices.length == 0 || prices == null) return 0;
        int profit = 0;
        int min = prices[0];
        int n = prices.length;

        
        for(int i = 1; i < n; i++) {
            min = Math.min(prices[i], min);
            profit = Math.max(prices[i] - min, profit);
        }
        
        return profit;
    }
    
}
